numpy/lib/recfunctions.py

```diff
@@ -70,6 +70,42 @@ def recursive_fill_fields(input, output):
     return output
 
 
+def get_fieldspec(dtype):
+    """
+    Produce a list of name/dtype pairs corresponding to the dtype fields
+
+    Similar to dtype.descr, but the second item of each tuple is a dtype, not a
+    string. As a result, this handles subarray dtypes
+
+    Can be passed to the dtype constructor to reconstruct the dtype, noting that
+    this (deliberately) discards field offsets.
+
+    Examples
+    --------
+    >>> dt = np.dtype([(('a', 'A'), int), ('b', float, 3)])
+    >>> dt.descr
+    [(('a', 'A'), '<i4'), ('b', '<f8', (3,))]
+    >>> get_fieldspec(dt)
+    [(('a', 'A'), dtype('int32')), ('b', dtype(('<f8', (3,))))]
+
+    """
+    if dtype.names is None:
+        # .descr returns a nameless field, so we should too
+        return [('', dtype)]
```

Comment:
Empty fields names causes headaches and bugs in the npy.io code, which I hope to fix once other PRs get in. "aligned" dtypes are particularly a problem, because `np.dtype('i1,f4', align=True).descr` is [('f0', '|i1'), ('', '|V3'), ('f1', '<f4')], so you often end up with lots of dummy fields named ''.

The point of my comment here is to say that actually your code is fine and I think you are probably fixing even more bugs than you think here, for aligned dtypes, because this function now skips padding bytes.